[DOC-12880] Auto Update Statistics

rakhi-prathap opened the pull request:
Documentation for a new Auto Update Statistics (AUS) feature for 8.0.

Preview links: 
- [Auto Update Statistics main page](https://preview.docs-test.couchbase.com/docs-devex-DOC-12880-auto-update-statistics/server/current/n1ql/n1ql-language-reference/auto-update-statistics.html)
- Other pages
     - [Advanced Query Features](https://preview.docs-test.couchbase.com/docs-devex-DOC-12880-auto-update-statistics/server/current/n1ql/advanced.html)
     - [Cost Based Optimizer](https://preview.docs-test.couchbase.com/docs-devex-DOC-12880-auto-update-statistics/server/current/n1ql/n1ql-language-reference/cost-based-optimizer.html)
     - [Using Cost Based Optimizer](https://preview.docs-test.couchbase.com/docs-devex-DOC-12880-auto-update-statistics/server/current/guides/cbo.html)
     - [Get System Information](https://preview.docs-test.couchbase.com/docs-devex-DOC-12880-auto-update-statistics/server/current/n1ql/n1ql-intro/sysinfo.html)

Preview credentials: [Preview docs for internal review](https://couchbasecloud.atlassian.net/wiki/x/dYF_dQ#Preview-docs-for-internal-review)

rakhi-prathap committed:
AUS rough  draft

rakhi-prathap committed:
Update system info table to add aus catalogs

rakhi-prathap committed:
Rough draft

rakhi-prathap committed:
Rewrite intro

rakhi-prathap committed:
Update content

rakhi-prathap committed:
Update content

rakhi-prathap committed:
Update content

rakhi-prathap committed:
Update content

rakhi-prathap committed:
Add PlantUML diagram

rakhi-prathap committed:
Update CBO pages

rakhi-prathap committed:
Update HEAD.yml for preview

rakhi-prathap committed:
Add a few more changes

rakhi-prathap committed:
Add a few more changes

rakhi-prathap committed:
Add a new section on canceling next scheduled AUS tasks

rakhi-prathap committed:
Add a new global setting, create_missing_statistics

rakhi-prathap committed:
Some formatting and rearrangment

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/auto-update-statistics.adoc:
```diff
@@ -0,0 +1,546 @@
+=== Update Phase
+
+After the evaluation, AUS executes xref:n1ql:n1ql-language-reference/updatestatistics.adoc[] statements to refresh the statistics identified as stale. 
+When updating the exisitng statistics, AUS ensures that the refreshed statistics maintain the original xref:n1ql:n1ql-language-reference/cost-based-optimizer.adoc#resolution[resolution] at which they were collected.
```
Minor: Typo in exisitng :-)

rakhi-prathap replied:
Sorry, fixed it.

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/auto-update-statistics.adoc:
```diff
@@ -0,0 +1,546 @@
+| <<aus_schedule, Schedule>> object
+
+| **change_percentage** +
+__optional__
```
Can we please change this from "optional" to "required" for change_percentage in system:aus?

My apologies, I accidentally made a mistake in the DOC ticket. Thank you!

rakhi-prathap replied:
No problem, changed it to "required".

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/auto-update-statistics.adoc:
```diff
@@ -0,0 +1,546 @@
+.Query
+[source,sqlpp]
+----
+INSERT INTO system:aus_settings ( KEY, VALUE ) 
```
The query is syntactically incorrect. The fix:

```
INSERT INTO system:aus_settings ( KEY, VALUE )
    VALUES ( "default:bucket1.scope1", {"change_percentage": 20 } );
```

rakhi-prathap replied:
Fixed it.

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/auto-update-statistics.adoc:
```diff
@@ -0,0 +1,546 @@
+=== Cancel Next Scheduled AUS Tasks
+
+To cancel the next scheduled AUS task, temporarily modify the schedule in the `system:aus` catalog using an UPDATE statement. 
+This way you can skip a specific AUS run without changing the overall configuration.
```
> This way you can skip a specific AUS run without changing the overall configuration.

I think this should be removed or modified? Because to cancel the next scheduled AUS task, you need to modify system:aus.. which does imply changing the overall configuration. But only temporarily.

rakhi-prathap replied:
Replaced this line with another one which says users can revert the schedule to the original settings.

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/auto-update-statistics.adoc:
```diff
@@ -0,0 +1,546 @@
+
+To handle this, AUS executes a scheduled task on each query node in the cluster. 
+This task evaluates statistics based on expiration policies to identify outdated ones and then refreshes them by running the xref:n1ql:n1ql-language-reference/updatestatistics.adoc[] statement.
+It primarily maintains statistics for expressions based on index keys, and not for expressions on non-indexed fields.
```
It primarily maintains statistics for expressions based on index keys, and not for expressions on non-indexed fields.

Could you please add to this sentence the following information:

 AUS only maintains statistics for expressions on index keys. But only for those keys indexed in Plasma GSI indexes. And not MOI indexes.

(Plasma and MOI are two different storage options for indexes.)

rakhi-prathap replied:
Done.

dhanyagowrish commented on modules/guides/pages/cbo.adoc:
```diff
@@ -93,6 +93,10 @@ For more information and examples, refer to xref:n1ql:n1ql-manage/query-settings
 Before you can use the Cost-Based Optimizer with a query, you must first gather the statistics that it needs.
 The Query service automatically gathers statistics whenever an index is created or built, and you can update statistics at any time.
 
+You can also configure a scheduled task to automatically check and update statistics using xref:n1ql:n1ql-language-reference/auto-update-statistics.adoc[Auto Updates Statistics (AUS)].
```
> Auto Update**s** Statistics (AUS)

Minor: Should be Auto Update Statistics (AUS)

rakhi-prathap replied:
My bad! Fixed it.

dhanyagowrish commented on modules/guides/pages/cbo.adoc:
```diff
@@ -220,6 +224,7 @@ Explanation:
 Reference:
 
 * xref:n1ql:n1ql-language-reference/updatestatistics.adoc[UPDATE STATISTICS]
+* xref:n1ql:n1ql-language-reference/auto-update-statistics.adoc[Auto Updates Statistics]
```
> Auto Update**s** Statistics (AUS)

Minor: Should be Auto Update Statistics (AUS)

rakhi-prathap replied:
My bad! Fixed it.

rakhi-prathap committed:
engg feedback + language and typo fixes

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/auto-update-statistics.adoc:
```diff
@@ -0,0 +1,537 @@
+====
+A sample task entry for a successful AUS task on a query node:
+
+[source,sqlpp]
```
This example is JSON

```suggestion
[source,json]
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/auto-update-statistics.adoc:
```diff
@@ -0,0 +1,537 @@
+= Auto Update Statistics
+:description: Auto Update Statistics (AUS) automatically refreshes optimizer statistics, ensuring accurate and cost-effective query plans.
```
* Add a label to flag this as a new feature in this release.
* If this is an Enterprise only feature, flag that as well
* Also you might want to consider increasing the TOC depth for this page, but this is not essential. Up to you

```suggestion
:page-status: Couchbase Server 8.0
:page-edition: Enterprise Edition
:page-toclevels: 2
:description: Auto Update Statistics (AUS) automatically refreshes optimizer statistics, ensuring accurate and cost-effective query plans.
```

simon-dew reviewed (changes requested):
This is great! Only a few suggestions. One question, is this an Enterprise-level feature? Please check, because CBO itself is Enterprise-only

rakhi-prathap committed:
Apply suggestions from code review

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

rakhi-prathap committed:
Updated the availability section

rakhi-prathap commented:
> This is great! Only a few suggestions. One question, is this an Enterprise-level feature? Please check, because CBO itself is Enterprise-only

Thank you, @simon-dew! Yes, this is an Enterprise-only feature (checked with the query team).
I've updated the Availability section to clarify this, and also incorporated your other suggestions.

simon-dew reviewed (approved):
LGTM 👍
